File: src/main/java/ca/mcgill/science/tepid/client/PrinterMgmt.java
package ca.mcgill.science.tepid.client;

import java.io.IOException;
import java.util.Map;

/**
 * Logic for managing printers for various systems
 */
public interface PrinterMgmt {

    /**
     * Adds a printer of the given data
     *
     * @param queue     name
     * @param port      port
     * @param isDefault true if name matches default
     */
    default void addPrinter(String queue, String port, boolean isDefault) {
        try {
            addPrinterImpl(queue, port, isDefault);
        } catch (IOException | InterruptedException e) {
            System.err.println("Add printer error");
            e.printStackTrace();
        }
    }

    /**
     * Delete a printer of the given data
     *
     * @param queue name
     * @param port  port
     */
    default void deletePrinter(String queue, String port) {
        try {
            deletePrinterImpl(queue, port);
        } catch (IOException | InterruptedException e) {
            System.err.println("Delete printer error " + e.getMessage());
        }
    }

    /**
     * Called when {@link Main} first starts.
     * Should add printers and remove them upon shutdown
     *
     * @param queueIds     map of ids
     * @param defaultQueue default queue id
     */
    default void bind(Map<String, String> queueIds, String defaultQueue) {
        if (!preBind()) {
            System.err.println("Failed to preBind printer management");
            return;
        }
        // add all printers
        queueIds.forEach((port, queue) -> addPrinter(queue, port, queue.equals(defaultQueue)));
        // delete all printers on shutdown
        Runtime.getRuntime().addShutdownHook(new Thread("Unmount Printers") {
            @Override
            public void run() {
                System.out.println("Unmounting " + queueIds.size() + " printers");
                queueIds.forEach((port, queue) -> deletePrinter(queue, port));
            }
        });
    }

    /**
     * Handler to setup content when binding is requested
     *
     * @return {@code true} if successful, {@code false} otherwise
     */
    boolean preBind();

    /**
     * Implicit implementation of {@link #addPrinter(String, String, boolean)} without the try catch
     */
    void addPrinterImpl(String queue, String port, boolean isDefault) throws IOException, InterruptedException;

    /**
     * Implicit implementation of {@link #deletePrinter(String, String)} without the try catch
     */
    void deletePrinterImpl(String queue, String port) throws IOException, InterruptedException;

    /**
     * Attempts to clean up the printers
     */
    void cleanPrinters();

    /**
     * @return filepath of tepid data
     */
    String tepidDataPath();

    /**
     * @return the appropriate {@link PrinterMgmt} for the current system
     */
    static PrinterMgmt getPrinterManagement() {
        return System.getProperty("os.name").startsWith("Windows") ? new WindowsPrinterMgmt() : new CupsPrinterMgmt();
    }


}

File: src/main/java/ca/mcgill/science/tepid/clientkt/printers/PrinterMgmt.kt
package ca.mcgill.science.tepid.clientkt.printers

import ca.mcgill.science.tepid.clientkt.utils.Config
import ca.mcgill.science.tepid.utils.WithLogging
import java.io.IOException

/**
 * Logic for managing printers for various systems
 */
interface PrinterMgmt {

    /**
     * Adds a printer of the given data
     *
     * @param queue     name
     * @param id      id
     * @param isDefault true if name matches default
     */
    fun addPrinter(queue: String, id: String, isDefault: Boolean) {
        val fullQueue = "$queue-${Config.USER_NAME}, id $id"
        log.info("Adding printer $fullQueue")
        try {
            addPrinterImpl(fullQueue, id, isDefault)
        } catch (e: IOException) {
            log.error("Add printer error", e)
        }
    }

    /**
     * Delete a printer of the given data
     *
     * @param queue name
     * @param id  id
     */
    fun deletePrinter(queue: String, id: String) {
        val fullQueue = "$queue-${Config.USER_NAME}"
        log.info("Deleting printer $fullQueue, id $id")
        try {
            deletePrinterImpl(fullQueue, id)
        } catch (e: IOException) {
            log.error("Delete printer error", e)
        }
    }

    /**
     * Called when client first starts.
     * Should add printers and remove them upon shutdown
     *
     * @param queueIds     map of ids
     * @param defaultQueue default queue id
     */
    fun bind(queueIds: Map<String, String>, defaultQueue: String?) {
        if (!preBind()) {
            log.error("Failed to preBind printer management")
            return
        }
        // delete all printers on shutdown
        Runtime.getRuntime().addShutdownHook(Thread({
            log.info("Unmounting ${queueIds.size} printers")
            queueIds.forEach { id, queue -> deletePrinter(queue, id) }
        }, "Unmount Printers"))

        // add all printers
        queueIds.forEach { id, queue -> addPrinter(queue, id, queue == defaultQueue) }
    }

    /**
     * Handler to setup content when binding is requested
     *
     * @return `true` if successful, `false` otherwise
     */
    fun preBind(): Boolean

    /**
     * Implicit implementation of [addPrinter] without the try catch
     */
    fun addPrinterImpl(queue: String, id: String, isDefault: Boolean)

    /**
     * Implicit implementation of [deletePrinter] without the try catch
     */
    fun deletePrinterImpl(queue: String, id: String)

    /**
     * Attempts to clean up the printers
     */
    fun cleanPrinters()

    /**
     * @return filepath of tepid data
     */
    fun tepidDataPath(): String

    companion object : WithLogging() {

        /**
         * @return the appropriate [PrinterMgmt] for the current system
         */
        val printerManagement: PrinterMgmt by lazy { if (Config.IS_WINDOWS) WindowsPrinterMgmt() else CupsPrinterMgmt() }
    }

}
